### PR TITLE
Add distributeNoReleaseAvailable: callback

### PR DESCRIPTION
**Pull request by Jamminroot**
Things to consider before you submit the PR:

* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

This PR adds `distributeNoReleaseAvailable:` callback to `MSACDistributeDelegate`, alert ui to demo apps and unit tests for new SDK code.

## Related PRs or issues

[AB#84054](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/84054)

### CODE REVIEW

File: AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h [hunk 1]
```diff
@@ -28,6 +28,11 @@ NS_SWIFT_NAME(DistributeDelegate)
  */
 - (BOOL)distribute:(MSACDistribute *)distribute releaseAvailableWithDetails:(MSACReleaseDetails *)details;
 
+/**
+ * Callback method that will be called update check reports that there is no new release available for update.
```
Comment (DmitriyKirakosyan) [inline AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h:32]:
```suggestion
 * Callback method that will be called whenever update check reports that there is no new release available for update.
```

File: AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h [hunk 2]
```diff
@@ -28,6 +28,11 @@ NS_SWIFT_NAME(DistributeDelegate)
...
+ */
+- (void)onNoReleaseAvailable;
```
Comment (DmitriyKirakosyan) [inline AppCenterDistribute/AppCenterDistribute/MSACDistributeDelegate.h:34]: We should add the NSDistribute parameter here, like:
```
- (void)distributeNoReleaseAvailable:(MSACDistribute *)distribute;
```

File: AppCenterDistribute/AppCenterDistributeTests/MSACDistributeTests.m [hunk 3]
```diff
@@ -836,6 +836,101 @@ - (void)testCheckLatestReleaseRemoveKeysOnNonRecoverableError {
...
+  details.status = @"available";
+  id detailsMock = OCMPartialMock(details);
+  OCMStub([detailsMock isValid]).andReturn(YES);
+  id delegateMock = OCMProtocolMock(@protocol(MSACDistributeDelegate));
+  id distributeMock = OCMPartialMock(self.sut);
+
+  OCMStub([distributeMock isNewerVersion:detailsMock]).andReturn(NO);
+  OCMStub([delegateMock onNoReleaseAvailable]).andDo(^(__unused NSInvocation *invocation) {
+    [expectation fulfill];
+    invoked = YES;
```
Comment (DmitriyKirakosyan) [inline AppCenterDistribute/AppCenterDistributeTests/MSACDistributeTests.m:854]: no need an extra variable here, `[expectation fulfill];` is enough.

File: AppCenterDistribute/AppCenterDistributeTests/MSACDistributeTests.m [hunk 4]
```diff
@@ -836,6 +836,101 @@ - (void)testCheckLatestReleaseRemoveKeysOnNonRecoverableError {
...
+        NSString *filename = [[NSBundle bundleForClass:[self class]] pathForResource:@"error_details_no_release" ofType:@"json"];
+        NSData *data = [NSData dataWithContentsOfFile:filename];
+        [invocation setArgument:&data atIndex:3];
+        [expectation fulfill];
+      })
+      .andForwardToRealObject();
+
+  OCMStub([delegateMock onNoReleaseAvailable]).andDo(^(__unused NSInvocation *invocation) {
+    [invokedExpectation fulfill];
+    invoked = YES;
```
Comment (DmitriyKirakosyan) [inline AppCenterDistribute/AppCenterDistributeTests/MSACDistributeTests.m:904]: same, redundant `invoked` var.

File: CHANGELOG.md [hunk 5]
```diff
@@ -144,7 +145,7 @@ ___
 ### App Center
 
 * **[Fix]** Optimization of release objects from memory during the execution of a large number of operations.
-* **[Fix]** Disable module debugging for release mode in the SDK to fix dSYM warnings.
+* **[Fix]** Disable module debugging for release mode in the SDK to fix dSYM warnings.
```
Comment (DmitriyKirakosyan) [inline CHANGELOG.md:148]: Please revert this line :)

File: Sasquatch/SasquatchObjC/AppDelegate.m [hunk 6]
```diff
@@ -381,6 +387,18 @@ - (BOOL)distribute:(MSACDistribute *)distribute releaseAvailableWithDetails:(MSA
   return NO;
 }
 
+- (void)onNoReleaseAvailable {
```
Comment (DmitriyKirakosyan) [inline Sasquatch/SasquatchObjC/AppDelegate.m:390]: We need to update SasquatchSwift as well.

File: Sasquatch/SasquatchObjC/AppDelegate.m [hunk 7]
```diff
@@ -381,6 +387,18 @@ - (BOOL)distribute:(MSACDistribute *)distribute releaseAvailableWithDetails:(MSA
...
+  NSLog(@"No release available callback invoked");
+  UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil
+                                                                 message:@"No updates available"
```
Comment (DmitriyKirakosyan) [inline Sasquatch/SasquatchObjC/AppDelegate.m:393]: I think the message should be localized.

File: AppCenterDistribute/AppCenterDistribute/MSACDistribute.m [hunk 8]
```diff
@@ -449,6 +449,16 @@ - (void)requestInstallInformationWith:(NSString *)releaseHash {
   }
 }
 
+- (void)checkDelegateAndInvokeOnNoReleaseAvailableCallback {
+  id<MSACDistributeDelegate> delegate = self.delegate;
+  if ([delegate respondsToSelector:@selector(onNoReleaseAvailableonNoReleaseAvailable)]) {
```
Comment (AnastasiaKubova) [inline AppCenterDistribute/AppCenterDistribute/MSACDistribute.m:454]: `onNoReleaseAvailableonNoReleaseAvailable` Is it typo?

Comment (Jamminroot) [reply AppCenterDistribute/AppCenterDistribute/MSACDistribute.m:454]: yup, it is 🤦‍♂️ Thanks

**Review by DmitriyKirakosyan — APPROVED** [review]

**Review by AnastasiaKubova — APPROVED** [review]

**Pull request merged**